Add React Labs March 2023 blog post (#5798)
reactjs/react.dev: src/content/blog/2023/03/22/react-labs-what-we-have-been-working-on-march-2023.md

---
title: "React Labs: What We've Been Working On – March 2023"
---

March 22, 2023 by [Joseph Savona](https://twitter.com/en_JS), [Josh Story](https://twitter.com/joshcstory), [Lauren Tan](https://twitter.com/potetotes), [Mengdi Chen](https://twitter.com/mengdi_en), [Samuel Susla](https://twitter.com/SamuelSusla), [Sathya Gunasekaran](https://twitter.com/_gsathya), [Sebastian Markbåge](https://twitter.com/sebmarkbage), and [Andrew Clark](https://twitter.com/acdlite)

---

<Intro>

In React Labs posts, we write about projects in active research and development. We've made significant progress on them since our [last update](https://react.dev/blog/2022/06/15/react-labs-what-we-have-been-working-on-june-2022), and we'd like to share what we learned.

</Intro>

---

## React Server Components {/*react-server-components*/}

React Server Components (or RSC) is a new application architecture designed by the React team.

> We've first shared our research on RSC in an [introductory talk](/blog/2020/12/21/data-fetching-with-react-server-components) and an [RFC](https://github.com/reactjs/rfcs/pull/188). To recap them, we are introducing a new kind of components--Server Components--that run ahead of time and are excluded from the browser JavaScript bundle. Server Components can run during the build, letting you read from the filesystem or fetch static content. They can also run on the server, letting you access your data layer without having to build an API. You can pass data by props from Server Components to the interactive Client components in the browser.

Author: ```suggestion
We've first shared our research on RSC in an [introductory talk](/blog/2020/12/21/data-fetching-with-react-server-components) and an [RFC](https://github.com/reactjs/rfcs/pull/188). To recap them, we are introducing a new kind of component--Server Components--that run ahead of time and are excluded from the browser JavaScript bundle. Server Components can run during the build, letting you read from the filesystem or fetch static content. They can also run on the server, letting you access your data layer without having to build an API. You can pass data by props from Server Components to the interactive Client components in the browser.
```

Today, people solve this problem with one of the two techniques.

One technique is to render a special third-party component that moves `<title>`, `<meta>`, and other tags inside it into the document `<head>`. This works for major browsers but there are many clients which do not run client-side JavaScript, such as Open Graph parsers, and so this technique is not universally suitable.

Another technique is to server-render the page in two parts. First, the main content is rendered and all such tags are collected. Then, the `<head>` is rendered with these tags. Finally, the `<head>` and the main content are sent to the browser. This approach works, but it prevents you from taking advantage of the [React 18's Streaming Server Renderer](/reference/react-dom/server/renderToReadableStream) because you'd have to wait for all content to render you can send the `<head>`.

> This is why we've created a solution that addresses all of these problems in a unified way. In short, tags like `<meta>`, `<link>`, and `<title>` anywhere in your component tree will be automatically deduplicated by React, preloaded during streaming, and inserted into the `<head>` automatically as they are discovered in the tree. This would work the same way in all environments, including fully client-side code, SSR, and in the future, RSC.

Reviewer: we no longer deduplicate document metadata, you still need to ensure you only render one title for instance otherwise there will be more than one title in the head.

Additionally most link types are not preloadable. We do selectively preload stylesheets that are not otherwise going to be hoisted but that's a very minor use case.

Finally, when streaming we don't guarantee every hoistable tag makes it into the head on the client b/c it's not actually all that useful. OG parsers can actually read the tags in the body for instance. Only stylesheets always hoist b/c their order is critical but then this isn't really document metadata.

Reviewer: ```suggestion
This is why we've created a solution that addresses all of these problems in a unified way. In short, tags like `<meta>`, `<link>`, and `<title>` anywhere in your component tree will be hoisted by React to the right place in the Document, preloaded if necessary during streaming, automatically as they are discovered in the tree. This will work the same way in all environments, including fully client-side code, SSR, and in the future, RSC.
```

## Transition Tracing {/*transition-tracing*/}

The Transition Tracing API lets you detect when React transitions become slower and to investigate why they may be slow. Following our last update, we have completed the initial design of the API and published an [RFC](https://github.com/reactjs/rfcs/pull/238). The basic capabilities have also been implemented. The project is currently on hold. We welcome feedback on the RFC and look forward to resuming its development to provide a better performance measurement tool for React transitions.
* * *
In addition to this update, our team has made recent guest appearances on community podcasts and livestreams to speak more on our work and answer questions.

* [Dan Abramov](https://twitter.com/dan_abramov) and [Joe Savona](https://twitter.com/en_JS) were interviewed by [Kent C. Dodds on his YouTube channel](https://www.youtube.com/watch?v=h7tur48JSaw), where they discussed concerns around React Server Components.
* [Dan Abramov](https://twitter.com/dan_abramov) and [Joe Savona](https://twitter.com/en_JS) were guests on the [JSParty podcast](https://jsparty.fm/267) and shared their thoughts about the future of React.

> Thanks to [Dave McCabe](https://twitter.com/mcc_abe), [Luna Wei](https://twitter.com/lunaleaps), [Matt Carroll](https://twitter.com/mattcarrollcode), [Sebastian Silbermann](https://twitter.com/sebsilbermann), [Seth Webster](https://twitter.com/sethwebster), and [Sophie Alpert](https://twitter.com/sophiebits) for reviewing this post.

Author: ```suggestion
Thanks to [Andrew Clark](https://twitter.com/acdlite), [Dan Abramov](https://twitter.com/dan_abramov), [Dave McCabe](https://twitter.com/mcc_abe), [Luna Wei](https://twitter.com/lunaleaps), [Matt Carroll](https://twitter.com/mattcarrollcode), [Sebastian Silbermann](https://twitter.com/sebsilbermann), [Seth Webster](https://twitter.com/sethwebster), and [Sophie Alpert](https://twitter.com/sophiebits) for reviewing this post.
```
